lib/async_storage.js

```diff
@@ -0,0 +1,103 @@
+'use strict';
+const { createHook, executionAsyncId } = require('async_hooks');
+const { WeakReference } = internalBinding('util');
+const { emitExperimentalWarning } = require('internal/util');
+
+emitExperimentalWarning('async_storage');
+// id => WeakReference<AsyncResource>
+const idResourcePairing = new Map();
+const storages = new Set();
+createHook({
+  init(asyncId, type, triggerAsyncId, resource) {
+    // save new resource
+    idResourcePairing.set(asyncId, new WeakReference(resource));
+    const triggerResourceWeak = idResourcePairing.get(triggerAsyncId);
+    if (triggerResourceWeak) { // it has a parent, let it be known
+      storages.forEach((storage) => {
+        storage._propagate(resource, triggerResourceWeak.get());
+      });
+    }
+  },
+  destroy(asyncId) {
+    idResourcePairing.delete(asyncId);
+  },
+  promiseResolve(asyncId) {
+    idResourcePairing.delete(asyncId);
+  }
+})
+  .enable();
+
+class AsyncStorage {
+
+  constructor() {
+    this.count = 0;
+    this.kstore = Symbol('async-storage');
```

Comment:
What if the resource object is not extensible? Does it make sense to use another WeakMap?

Also, please follow the [Symbol usage guidelines](https://github.com/nodejs/node/blob/master/doc/guides/using-symbols.md), e.g. use `this.kStore = Symbol('kStore')`.

Reply:
Indeed, I will move back to a WeakMap here